src/main/java/org/threadly/concurrent/event/ListenerHelper.java

```diff
@@ -146,16 +145,17 @@ public void addListener(T listener, Executor executor) {
       if (addingFromCallingThread) {
         // we must create a new instance of listeners to prevent a ConcurrentModificationException
         // we know at this point that listeners can not be null
-        Map<T, Executor> newListeners = new HashMap<T, Executor>(listeners.size() + 1);
-        newListeners.putAll(listeners);
-        newListeners.put(listener, executor);
+        List<Pair<T, Executor>> newListeners = 
+            new ArrayList<Pair<T, Executor>>(listeners.size() + 1);
```

Comment:
With all this copying of the array on add/remove, would it just be better to use a copy on write Array?  Seems like it would save us from having to do this logic.

Also maybe I am wrong here, but I seem to remember setting the size to the size we are going to be adding to the arrayList is just going to cause a resize on the arrayList once we add the last item.  I think you are supposed to add totalsize+2 to keep it from doing that.